### PR TITLE
Support for decorating resources

### PR DESCRIPTION
**Pull request by pustomytnyk**
Allows use of decorator pattern (e.g. [Draper](https://github.com/drapergem/draper)). One can define `#decorate_resource` on a dashboard.

- [x] Decoration of resource and collection (pagination friendly)

- [x] Add an example to example app

- [x] Tests for classes which use new methods

### CODE REVIEW

**Comment by nickcharlton** [comment]
Hi @pustomytnyk! This is looking great. Would you be able to introduce some tests?

**Comment by pustomytnyk** [comment]
ok. I'll add tests and feature to demo app. Not sure when.

**Comment by pustomytnyk** [comment]
done. Don't know what is the issue with failed test. I think for once it occurred locally too.

File: spec/example_app/app/decorators/order_decorator.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  def address
+    [address_city, address_state].join(', ')
+  end
+
+  def self.state_by_abbr(abbr)
+    index = STATE_ABBR.index(abbr)
+    return nil unless index
+    STATE[index]
```
Comment (BenMorganIO) [inline spec/example_app/app/decorators/order_decorator.rb]: Refactor suggestion:

```ruby
index = STATE_ABBR.index(abbr)
STATE[index] if index
```

File: spec/example_app/app/decorators/decorator_base.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class DecoratorBase
+  def initialize(model)
+    @model = model
+  end
+
+  def class
+    @model.class
+  end
+
+  def method_missing(method, *args, &block)
```
Comment (BenMorganIO) [inline spec/example_app/app/decorators/decorator_base.rb]: This scares me. What's the need for meta-programming here?

**Comment by pustomytnyk** [comment]
@BenMorganIO, I've replaced meta-delegating with extending.

**Comment by pustomytnyk** [comment]
@nickcharlton @BenMorganIO any suggestions?

**Comment by pustomytnyk** [comment]
Moved decorating to page presenters.

**Comment by pustomytnyk** [comment]
I'm going to add controversial change to the pull request in order to solve #911
Namely, to pass chain of rendering contexts (ruby hashes) to every Page and Field. This would allow 1) to have conditional rendering based on rendering context (as described in issue #911); 2) to pass controller variables to renderer (eg. user's timezone)

**Comment by pustomytnyk** [comment]
It still requires few fixes and tests at the moment.

**Comment by schovi** [comment]
This is awesome!

**Comment by pustomytnyk** [comment]
I've added tests for modified classes and a hint in the dashboard generator template.

**Comment by caitlinpb** [comment]
Any chance of this getting conflicts fixed and merged in? Could be very useful :)

**Comment by pustomytnyk** [comment]
@caitlinpb I didn't look to the gem and my PRs for a long time :) Not sure if I'll have time to check it till New Year.

**Comment by pablobm** [comment]
I'm going to close due to lack of activity 😞 If someone wants to take over, let us know!

**Comment by dgmstuart** [comment]
> I'm going to close due to lack of activity 😞 If someone wants to take over, let us know!

@pablobm I've taken this over in a new PR: https://github.com/thoughtbot/administrate/pull/1773

**Pull request closed without merging**